open-craft/ecommerce
# [BB-3801] HyperPay payment processor

# Do not merge to `opencraft-release/juniper.3` even after this PR is reviewed/approved.

## Description

This PR implements a new payment processor backend for HyperPay - https://hyperpay.docs.oppwa.com/tutorials/integration-guide#

## How it works

* The `ecommerce/extensions/payment/processors/hyperpay.py` file implements 2 separate ecommerce payment backends which are very closely related to each other - `HyperPay` and `HyperPayMada`. Two separate processors are needed because they use separate "entity IDs" (a value that is required by the HyperPay API) and both have to be enabled at the same time. So the `HyperPayMada` class inherits frorm the `HyperPay` class and just overrides some configuration and variables.
* For each enabled processor from this backend, there will be a button that shows up on the checkout page.
* On clicking one of the relevant checkout buttons, some JavaScript code executes and eventually, the control comes to the `get_transaction_parameters()` method in the backend.
* This method tries to create a checkout as detailed [here](https://hyperpay.docs.oppwa.com/integration-guide) in the HyperPay integration documentation.
* Then the user has to be redirected to a page with the appropriate JavaScript loaded which renders the payment form in an iframe and handles the payment processing from there.
* In the ecommerce server-side checkout process, ecommerrce will POST the transaction parameters returrned by the `get_transaction_parameters()` method to the `payment_page_url` URL returned by the same method. This payment page URL will be a page hosted on the payment provider's platform. But in this case, we have to load our own page with the custom JavaScript code embedded.
* So there is an internal view implemented for this page which accepts all the transaction parameters as POST data, along with the CSRF token and passes them on to a Django template which renders the payment page.
* The JavaScript code in that page can be used to customize various aspects of the payment form, fields etc. as documented [here](https://hyperpay.docs.oppwa.com/tutorials/integration-guide/customisation) and [here](https://hyperpay.docs.oppwa.com/tutorials/integration-guide/advanced-options). The billing details are collected in the payment form and not from the LMS user profile.
* In case of a client-side checkout, the backend doesn't do any of the work done above and everything has to be done in the frontend page and its JavaScript to submit the payment data to the payment provider's servers and handle the processing. But for HyperPay, our client-side responsibility stops with rendering the payment page and everything else is handled by the embedded iframe. This backend is a hybrid of the server-side and client-side checkout methods.
* So the user then enters the payment information and completes the payment flow. At the end of it, the HyperPay server redirects to a well-known return URL on the ecommerce app with the relevant transaction parameters via a GET request. This can be a POST request for some other payment processors.
* There has to be a Django view sub-classing `EdxOrderPlacementMixin` that handles this.
* The view determines the payment status endpoint URL from the GET data and sends a post request to that URL along with the entity ID that was originally used to create the checkout.
* Then the status of the transaction is determined and if successful, the user is redirected to the receipt page and to the error page otherwise.
* Irrespective of the status, errors, all the responses from the HyperPay API are recorded in the database for tracking purposes.
* Before redirecting the user to the receipt or the error page, the view calls the `handle_payment()` method, which in turn calls the `handle_processor_response()` method on the main backend class. Some details about the transaction like the transaction id, card number etc. are captured there.


## Testing instructions
* On a fresh juniper devstack, check out this PR's source branch.
* In `ecommerce/settings/devstack.py`, add the following configuration to the `PAYMENT_PROCESSOR_CONFIG` dictionary for the `edX` partner.

  ```code:python
        'hyperpay': {
            'access_token': '<access-token>',
            'entity_id': '<entity-id for visa/mastercard>',
            'currency': 'SAR',
            'hyperpay_base_api_url': 'https://test.oppwa.com',
            'return_url': '/payment/hyperpay/submit/'
        },
        'hyperpay_mada': {
            'access_token': '<access-token>',
            'entity_id': '<entity-id for mada>,
            'currency': 'SAR',
            'hyperpay_base_api_url': 'https://test.oppwa.com',
            'return_url': '/payment/hyperpay/mada/submit/'
        },
   ```
* Restart the ecommerce service.
* Set `hyperpay,hyperpay_mada` as the payment processors for the default ecommerce site in its site configuration.
* Set the `enable_client_side_checkout` waffle flag in the ecommerce admin site to `No` for everyone.
* Add the waffle switches `payment_processor_active_hyperpay`, `payment_processor_active_hyperpay_mada` and activate them.
* Log in to the LMS as a new user.
* Enroll in the demo course if not already enrolled.
* Click on the button to start the payment and verification process.
* In the cart page, verify that there are two buttons, one for credit card and one for mada.
* Start the payment process and complete it successfully.  Use the test card details shared on the task.
* Verify that the payment page is themed like all ecommerce pages.
* Verify that the billing details are required in the payment form.
* Verify that the CVV field is masked
* Verify that the enrollment is successful and the receipt page is shown.
* Start the payment process again for a different user and fail it. Verify that the enrollment didn't succeed and the error page is shown.
* Log in to the ecommerce admin site and verify that all the responses from HyperPay are recorded in the Payment processor response model.
* Repeat the above steps for both the credit card and mada payment methods. Use the test card details shared on the task.
* Run the unit tests and verify that they pass.
* Do a code review.

## Author notes and concerns
* Should the locale of the payment form be blindly set to `settings.LANGUAGE_CODE` since there is nothing like the dark lang middleware in ecommerce? Or should it try to determine the locale from the browser headers and use it instead?
* There are a minimal number of tests written. A lot more can be written if we have the time.
* This doesn't yet use the pluggable backend feature that we implemented recently for another client.
